### ISSUE TITLE
Bug - Gmail.js not defined in "new message" only window

### ISSUE THREAD

**Issue by JerryBels** [KartikTalwar/gmail.js#194]
Hello !

So I found a bug that's really annoying. If someone ctrl+clic on the "new message" button of gmail, it opens a new tab containing the form to send the mail.

But, in that page, Gmail.js doesn't work. In my own plugin I'm working on, it says "GLOBALS is not defined", and even the boilerplate doesn't work at all, showing "Gmail is not defined".

To reproduce, simply : 
1/ Grab a fresh install of https://github.com/KartikTalwar/gmail-chrome-extension-boilerplate
2/ Add to it latest https://github.com/KartikTalwar/gmail.js/blob/master/src/gmail.js
3/ Add it to Chrome
4/ Open your Gmail tab. The console shows "Hello xxx@gmail.com" as expected.
5/ Ctrl+clic on the "new message" big red button
6/ On the newly opened tab, open console to see the error message, and no hello :(

**Comment by KartikTalwar** [comment]
Hey @JerryBels, this is expected since `GLOBALS` is the js variable that stores a few key user info in gmail.com which this library accesses. A popup from either a compose/print/etc does not contain the variable therefore the initialization fails. Would it be possible for you to describe what your extension wants to do in the popup window? There are a few things that you'd definitely not be able to do but I can look into adding an override

**Comment by JerryBels** [comment]
Hey @KartikTalwar, thanks for answering !

Well, we're doing many things in our extension, using Gmail.js, that we need when a new compose window is opened. Detecting the user's email is the first. Then : 

    gmail.observe.on("compose")

is used to inject a toolbar.

gmail.observe.before('send_message')

is used to add elements to the body of the message.

provider.observe.after('send_message')

is used to get the ID of the mail sent and send it over Ajax to our server.

Currently that's all we do, but I guess that would be great to have this tab working like every other, although I realise now this could be a bit tricky since Gmail doesn't work the same in that tab, since it's more like a popup...

Thanks again !
Jeremy

**Comment by KartikTalwar** [comment]
@JerryBels upon closer inspection, if you send a message from the popup window, the action still gets transferred back to the main gmail window which makes the request. This means that the observer will fire in the main window and should not require re-initialization in the popup window.

Because those `compose` and `send_message` observers don't require any `GLOBALS` properties, can you try changing line 27-31 in your gmail.js to these. Hopefully this will make everything work

```js
  api.version           = "0.4";
  api.tracker.globals   = typeof GLOBALS !== "undefined" ? GLOBALS: [];
  api.tracker.view_data = typeof VIEW_DATA !== 'undefined' ? VIEW_DATA : [];
  api.tracker.ik        = api.tracker.globals[9] || "";
  api.tracker.hangouts  = undefined;```

**Comment by JerryBels** [comment]
Hey @KartikTalwar,

I did that test, interesting idea, but...

First, of course the current email is still undefined (hello, undefined !)

Then, also, there is a new error about this line in xhr_watcher : 

    var win = top.document.getElementById("js_frame").contentDocument.defaultView;

The error : 

    Uncaught TypeError: Cannot read property 'contentDocument' of null

Of course the js_frame doesn't exist in this environment.

About that, you keep refering about it as "the popup". I don't know if it's a name just for easy reference, but for me it doesn't open a real popup. Hitting that ctrl+clic opens a new browser tab.

1 more interesting thing : I don't know if you know but gmail can be defined in Chrome to natively open mailto links. When it's the case, the tab opened is very much the same as the one that shows up when hitting the new message button on gmail with ctrl+clic.... Except everything works exactly as intended. GLOBALS is defined, etc... I don't know if it can help.

**Comment by JerryBels** [comment]
Okay, made some progress. I can get rid of the error by changing xhr_watcher function from this : 

    var win = top.document.getElementById("js_frame").contentDocument.defaultView;

to this : 

    var win = top.document.getElementById("js_frame") ? top.document.getElementById("js_frame").contentDocument.defaultView : window.opener.top.document.getElementById("js_frame").contentDocument.defaultView;

No more errors. BUT ! The injected html still doesn't shows up.

I simply added a console.log in my code, where I call

    Gmail.observe.on("compose", function(compose, type) {

It works when a regular new message is created, but when it's the popup, nothing. The event is simply not fired at all.

Do you have any idea of what could be wrong @KartikTalwar ?

**Comment by KartikTalwar** [comment]
As a clarification, when I say popup I do mean the new tab that opens. I am currently traveling but I can have a look at this on Tuesday when I get back. Are those the only changes you made? Can you push them in your fork of gmail.js?

**Comment by JerryBels** [comment]
Sure : https://github.com/JerryBels/gmail.js

**Comment by JerryBels** [comment]
@KartikTalwar , hey man, did you have time to take a little look at this issue ?

**Comment by JerryBels** [comment]
Hey !

I did solve the issue. Actually, any code placed in 

    Gmail.observe.on("compose", function(compose, type) {

will work for NEWLY opened message windows. Windows that are already opened BEFORE gmail.js loads won't trigger the event - thus, nothing will happen.

To fix that in my code, I simply used 

    var openedComposes = provider.dom.composes();
	for(var i = 0; i < openedComposes.length; i++) {
		// do things, like calling the same function Gmail.observe.on("compose") calls
	}

And everything is fine :)

I will open a pull request with my changes to gmail.js, if you want to merge them in the main src for others to not bump into this problem.
